apps/networkmonitor/networkmonitor.nim

```diff
@@ -82,43 +99,68 @@ proc setConnectedPeersMetrics(discoveredNodes: seq[Node],
+          return ok()
+
+        except CatchableError:
+          var msg = getCurrentExceptionMsg()
+          if msg == "Future operation cancelled!":
+            msg = "timedout"
+          warn "failed to ping the peer", peer=peerInfo, err=msg
+
+          allPeers[peerId].connError = msg
+          return err("could not ping peer: " & msg)
+
+      let timedOut = not await ping().withTimeout(timeout)
+      # need this check for pingDelat == 0 because there may be a conn error before timeout
+      if timedOut or pingDelay == 0.millis:
+        allPeers[peerId].retries += 1
+        continue
+
+      info "successfully pinged peer", peer=peerInfo, duration=pingDelay.millis
+      peer_ping.observe(pingDelay.millis)
+
+      if allPeers[peerId].avgPingDuration == 0.millis:
+        allPeers[peerId].avgPingDuration = pingDelay
+
+      # TODO: check why the calculation ends up losing precision
+      allPeers[peerId].avgPingDuration = int64((float64(allPeers[peerId].avgPingDuration.millis) * (AvgPingWindow - 1.0) + float64(pingDelay.millis)) / AvgPingWindow).millis
+      allPeers[peerId].lastPingDuration = pingDelay
 
-    # after connection, get supported protocols
-    let lp2pPeerStore = node.switch.peerStore
-    let nodeProtocols = lp2pPeerStore[ProtoBook][peer.get().peerId]
-    allPeers[peerId].supportedProtocols = nodeProtocols
-    allPeers[peerId].lastTimeConnected = currentTime
+      # after connection, get supported protocols
+      let lp2pPeerStore = node.switch.peerStore
+      let nodeProtocols = lp2pPeerStore[ProtoBook][peerInfo.peerId]
+      allPeers[peerId].supportedProtocols = nodeProtocols
+      allPeers[peerId].lastTimeConnected = currentTime
 
-    # after connection, get user-agent
-    let nodeUserAgent = lp2pPeerStore[AgentBook][peer.get().peerId]
-    allPeers[peerId].userAgent = nodeUserAgent
+      # after connection, get user-agent
+      let nodeUserAgent = lp2pPeerStore[AgentBook][peerInfo.peerId]
+      allPeers[peerId].userAgent = nodeUserAgent
 
-    # store avaiable protocols in the network
-    for protocol in nodeProtocols:
-      if not allProtocols.hasKey(protocol):
-        allProtocols[protocol] = 0
-      allProtocols[protocol] += 1
+      # store avaiable protocols in the network
+      for protocol in nodeProtocols:
+        if not allProtocols.hasKey(protocol):
+          allProtocols[protocol] = 0
+        allProtocols[protocol] += 1
 
-    # store available user-agents in the network
-    if not allAgentStrings.hasKey(nodeUserAgent):
-      allAgentStrings[nodeUserAgent] = 0
-    allAgentStrings[nodeUserAgent] += 1
+      # store available user-agents in the network
+      if not allAgentStrings.hasKey(nodeUserAgent):
+        allAgentStrings[nodeUserAgent] = 0
+      allAgentStrings[nodeUserAgent] += 1
 
-    debug "connected to peer", peer=allPeers[customPeerInfo.peerId]
+      debug "connected to peer", peer=allPeers[customPeerInfo.peerId]
```

Comment:
Afaict this extra indentation will now result in these being only set if the ping condition is met. Perhaps intentional?